dbms/src/Interpreters/InterpreterDropQuery.cpp

```diff
@@ -34,161 +36,170 @@ BlockIO InterpreterDropQuery::execute()
-    if (!database && !drop.if_exists)
-        throw Exception("Database " + database_name + " doesn't exist", ErrorCodes::UNKNOWN_DATABASE);
-
-    std::vector<std::pair<StoragePtr, std::unique_ptr<DDLGuard>>> tables_to_drop;
-
-    if (!drop_database)
-    {
-        StoragePtr table;
+    String database_name = database_name_.empty() ? context.getCurrentDatabase() : database_name_;
 
-        if (drop.if_exists)
-            table = context.tryGetTable(database_name, drop.table);
-        else
-            table = context.getTable(database_name, drop.table);
+    DatabaseAndTable database_and_table = tryGetDatabaseAndTable(database_name, table_name, if_exists);
 
-        if (table)
-            tables_to_drop.emplace_back(table,
-                context.getDDLGuard(
-                    database_name, drop.table, "Table " + database_name + "." + drop.table + " is dropping or detaching right now"));
-        else
-            return {};
-    }
-    else
+    if (database_and_table.first && database_and_table.second)
     {
-        if (!database)
-        {
-            if (!drop.if_exists)
-                throw Exception("Database " + database_name + " doesn't exist", ErrorCodes::UNKNOWN_DATABASE);
-            return {};
-        }
+        auto ddl_guard = context.getDDLGuard(
+            database_name, table_name, "Table " + database_name + "." + table_name + " is dropping or detaching right now");
 
-        for (auto iterator = database->getIterator(context); iterator->isValid(); iterator->next())
-            tables_to_drop.emplace_back(iterator->table(),
-                context.getDDLGuard(database_name,
-                    iterator->name(),
-                    "Table " + database_name + "." + iterator->name() + " is dropping or detaching right now"));
-    }
-
-    for (auto & table : tables_to_drop)
-    {
-        if (!drop.detach)
+        if (kind == ASTDropQuery::Kind::Detach)
         {
-            if (!table.first->checkTableCanBeDropped())
-                throw Exception("Table " + database_name + "." + table.first->getTableName() + " couldn't be dropped due to failed pre-drop check",
-                    ErrorCodes::TABLE_WAS_NOT_DROPPED);
+            database_and_table.second->shutdown();
+            /// If table was already dropped by anyone, an exception will be thrown
+            auto table_lock = database_and_table.second->lockDataForAlter(__PRETTY_FUNCTION__);
+            /// Drop table from memory, don't touch data and metadata
+            database_and_table.first->detachTable(database_and_table.second->getTableName());
         }
-
-        table.first->shutdown();
-
-        /// If table was already dropped by anyone, an exception will be thrown
-        auto table_lock = table.first->lockForAlter(__PRETTY_FUNCTION__);
-
-        String current_table_name = table.first->getTableName();
-
-        if (drop.detach)
+        else if (kind == ASTDropQuery::Kind::Truncate)
```

Comment:
`checkTableCanBeDropped` should be called for truncate, because it is equally dangerous operation.